ledger/sandbox/src/test/lib/scala/com/digitalasset/platform/store/dao/JdbcLedgerDaoCompletionsSpec.scala

```diff
@@ -0,0 +1,147 @@
+  private val parties: Set[Party] = Set(party)
+
+  private def rejectWith(reason: RejectionReason): PersistenceEntry.Rejection =
+    PersistenceEntry.Rejection(
+      LedgerEntry.Rejection(
+        recordTime = Instant.now,
+        commandId = UUID.randomUUID().toString,
+        applicationId = applicationId,
+        submitter = party,
+        rejectionReason = reason,
+      )
+    )
+
+  it should "return the expected completion for a rejection" in {
+    val offset = nextOffset()
+    val rejection = rejectWith(RejectionReason.Inconsistent(""))
+    for {
+      from <- ledgerDao.lookupLedgerEnd()
+      _ <- ledgerDao.storeLedgerEntry(offset, rejection)
+      to <- ledgerDao.lookupLedgerEnd()
+      (_, response) <- ledgerDao.completions
+        .getCommandCompletions(from, to, applicationId, parties)
+        .runWith(Sink.head)
+    } yield {
+      val receivedOffset =
+        ApiOffset.assertFromString(response.checkpoint.value.offset.value.value.absolute.value)
+      receivedOffset shouldBe offset
+
+      response.completions should have length 1
+      val completion = response.completions.head
+
+      completion.transactionId shouldBe empty
+      completion.commandId shouldBe rejection.entry.commandId
+      completion.status.value.code shouldNot be(io.grpc.Status.Code.OK)
+    }
+  }
+
+  it should "not return completions if the application id is wrong" in {
+    val offset = nextOffset()
+    val rejection = rejectWith(RejectionReason.Inconsistent(""))
+    for {
+      from <- ledgerDao.lookupLedgerEnd()
+      _ <- ledgerDao.storeLedgerEntry(offset, rejection)
+      to <- ledgerDao.lookupLedgerEnd()
+      response <- ledgerDao.completions
+        .getCommandCompletions(from, to, applicationId = "WRONG", parties)
+        .runWith(Sink.seq)
+    } yield {
+      response should have length 0
+    }
+  }
+
+  it should "not return completions if the parties do not match" in {
+    val offset = nextOffset()
+    val rejection = rejectWith(RejectionReason.Inconsistent(""))
+    for {
+      from <- ledgerDao.lookupLedgerEnd()
+      _ <- ledgerDao.storeLedgerEntry(offset, rejection)
+      to <- ledgerDao.lookupLedgerEnd()
+      response <- ledgerDao.completions
+        .getCommandCompletions(from, to, applicationId, Set("WRONG"))
+        .runWith(Sink.seq)
+    } yield {
+      response should have length 0
+    }
+  }
+
+  it should "return the expected status for each rejection reason" in {
+    val reasons = Seq[RejectionReason](
+      RejectionReason.Disputed(""),
+      RejectionReason.Inconsistent(""),
+      RejectionReason.InvalidLedgerTime(""),
+      RejectionReason.OutOfQuota(""),
+      RejectionReason.PartyNotKnownOnLedger(""),
+      RejectionReason.SubmitterCannotActViaParticipant(""),
+      RejectionReason.TimedOut(""),
+    )
+
+    for {
+      from <- ledgerDao.lookupLedgerEnd()
+      _ <- Future.sequence(
+        for (reason <- reasons; offset = nextOffset())
```

Comment:
You can probably inline `offset`. Even if you don't, I think it'd be clearer if it lived in the `yield` block.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/5072/commits/a18bf358d4c7d743d9738aee6fb84c70da1d0afd